### DISCUSSION TITLE
What takes precedence when using `brew install <token>`? Formulas from taps or casks from the centralised repo?

### DISCUSSION THREAD

**Discussion by AndyTitu** [Homebrew/discussions#4693] [Everyday usage]
### Output of `brew config`

```shell
-
```

### Output of `brew doctor`

```shell
-
```

### Description of issue

I am mainly interested in determining the order of precedence for the place brew will pull the app from.

With the latest version of brew (on MacOS), brew install <token> will automatically detect if the app which was identified with the provided token is a cask or formula, and install the appropriate one.

If we are only looking at homebrew-core and homebrew-cask for the sources of formula/casks, brew will first check if the given token matches an existing formula and, if not, it will check for casks.

The question becomes tricky when taps come into play. I would then expect the order of precedence to be:

first check formulas from core against the token
then check formulas from tapped taps against the token
then check casks from homebrew/cask against the token
then check casks from the tapped taps against the token
Is this already the case? Or is there a different order in place?

**Comment by Bo98** [comment]
That order sounds correct, yes.

If you know precisely what you want, I generally recommend using `--formula` or `--cask`, and to specify the full name if installing from a third-party tap (e.g. `some/tap/formula`), given we offer no guarantees that a formula with the same name as a cask won't appear in the future, and we don't offer guarantees on the order of searching taps (except that homebrew/core or homebrew/cask is first) if two third-party taps happened to have formulae or casks of the same name.

**Reply by apainintheneck** [reply]
I agree that order sounds correct.

To add to this, there are some safe guards around handling the ambiguity of multiple formulae/casks with the same name (though they should be unique per tap).

If we default to a formula and `--formula` is not set, we check to see if there is a cask with the same name and provide a warning in that case (the install continues though).

Beyond that, if a formula or cask is not available from one of the core repos and there are multiple third-party taps that have packages with the same searched for name, we throw an error explaining the ambiguity.
